### DISCUSSION TITLE
Concepts for development

### DISCUSSION THREAD

**Discussion by TecDroiD** [lvgl-micropython/lvgl_micropython#30] [General]
Hi there,
I'm relatively new to lvgl (well.. that's a lie. Just doing my first project) and have some questions about concepts behind lvgl and especially the mp binding.. Maybe discussing them here helps me- and others, too

This is a multi-question-thread - sorry for that..

1. From what I can see, there is no direct connection between the display I use and lvgl. I initialize the display and then create a screen out of nowhere. Would have expected some kind of scrn = lv.screen_on_display(myLCD). So how does that thing know where to put my data?
2. Is there any order of doing things? I mean, initializing the display driver, doing lv.init(), etc.
3. How do people design their program? I thought about creating the display in boot.py so I can concentrate on whatever my sofware will do in main.py (and other modules)
4. are there any good tutorials on the basics?
5. Does someone know of any wysiwyg editors for lvgl capable of generating mp code yet? I thought about creating some json based system for building views... but that's really much work to do I think..

**Comment by kdschlosser** [comment]
>From what I can see, there is no direct connection between the display I use and lvgl. I initialize the display and then create a screen out of nowhere. Would have expected some kind of scrn = lv.screen_on_display(myLCD). So how does that thing know where to put my data?

It's doesn't come out of "nowhere". The mechanics are all behind the scenes and not out in plain view.

There are 2 pieces needed to get a display to work. The first one is the bus.the bus is the physical connection from the display to the MCU. This is typically I8080, SPI or RGB. There are others but at the moment that is what this binding supports. Those are also the most common ones used for DIY'ers.

The bus drivers are written in C code and they are written in a manner where the functions have the same API . There is a reason why this is important and I will explain in a little bit.

The second part is the driver IC for the display. This IC is responsible for communicating with the MCU and for sending data to the raw display. There are many different driver ICs available. I have included the software portion of the driver IC's for the most commonly used ones. 

The reason why it is important for the functions to be the same across the different busses is because a  driver IC might support more than one type of bus. It can only use one at a time but there is the option there to choose which one is going to be used. Because of the the software that handles the display needs to have zero knowledge of what kind of bus is being used, it is only going to make the code more complex if things are different between the different busses.

The software that handles the driver ICs  is written in Python. 

>Is there any order of doing things? I mean, initializing the display driver, doing lv.init(), etc.

That all depends on the MCU being used. 

>How do people design their program? I thought about creating the display in boot.py so I can concentrate on whatever my sofware will do in main.py (and other modules)

boot.py would be used for things like setting up the WiFi or Bluetooth. The main.py is where you put all of the application code including the code needed to set up the display. The reason why is because of how micropython works and being able to flash the MCU and also with uploading files.

>are there any good tutorials on the basics?

good question

>Does someone know of any wysiwyg editors for lvgl capable of generating mp code yet? I thought about creating some json based system for building views... but that's really much work to do I think..

Squareline Studio is the only one that I am aware of.
